### PR TITLE
Resolve duplication

### PR DESCRIPTION
**Pull request by beliaev-maksim**
wrap duplicated code fragments into functions

### CODE REVIEW

File: tests/unit/test_endpoint_provider.py [hunk 1]
```diff
@@ -278,36 +278,32 @@ def test_provider_unit_sets_fqdn_if_not_address_on_relation_joined(self):
         self.assertEqual(data["prometheus_scrape_unit_address"], "some.host")
         self.assertIn("prometheus_scrape_unit_name", data)
 
-    @patch_network_get()
-    def test_provider_supports_multiple_jobs(self):
+    def _get_data(self, target):
```
Comment (sed-i) [inline tests/unit/test_endpoint_provider.py:281]: Can we generalize this a function?
Seems odd to me that"
- `provider` is hard-coded
- a non-test function is making assertions
- this specifically addresses app data (cf. https://github.com/canonical/operator/issues/834)

On one hand it reduces duplication but on the other hand the tests become less straightforward.

**Review by sed-i — COMMENTED** [review]
Most changes look good.
I have reservations about `_get_data`.

File: tests/unit/test_endpoint_provider.py [hunk 2]
```diff
@@ -585,25 +574,30 @@ def gen_group(self, name):
     def test_load_multiple_rules_per_file(self):
         """Test official format with multiple alert rules per group in multiple groups."""
         rules_file_dict = {"groups": [self.gen_group(1), self.gen_group(2)]}
+        self._compare_rules(rules_file_dict, "1", "2")
+
+    def _generate_rules_file_dict(self, rules_file_dict):
```
Comment (dstathis) [inline tests/unit/test_endpoint_provider.py:579]: This method seems to take a dict, write it to disk, read it from disk, and return it. What is the difference between this and
```
def generate_rules_file_dict(self, rules_file_dict):
    return rules_file_dict
```

**Review by dstathis — COMMENTED** [review]
Unfortunately, I think this PR sacrifice a lot of readability for a small amount of code deduplication.

File: tests/unit/test_endpoint_provider.py (hunk 2 above)
Comment (beliaev-maksim) [reply tests/unit/test_endpoint_provider.py:579]: checks that alert rules are properly reading it?

PS
it is not a code that I wrote initially :)

Comment (beliaev-maksim) [reply tests/unit/test_endpoint_provider.py:579]: also, I think it will insert juju topology in the resulting dict

**Pull request closed without merging**